dhis-2/dhis-api/src/main/java/org/hisp/dhis/keyjsonvalue/KeyJsonValueQuery.java

```diff
@@ -0,0 +1,203 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.keyjsonvalue;
+
+import static java.lang.Character.isLetterOrDigit;
+import static java.util.Collections.emptyList;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.ToString;
+
+import org.hisp.dhis.common.NamedParams;
+
+/**
+ * Details of a query as it can be performed to fetch {@link KeyJsonValue}s.
+ *
+ * @author Jan Bernitt
+ */
+@ToString
+@Getter
+@Builder( toBuilder = true )
+@AllArgsConstructor( access = AccessLevel.PRIVATE )
+public final class KeyJsonValueQuery
+{
+
+    private final String namespace;
+
+    /**
+     * By default, only entries which have at least one non-null value for the
+     * extracted fields are returned. If all are included even matches with only
+     * null values are included in the result list.
+     */
+    private final boolean includeAll;
+
+    @Builder.Default
+    private final List<Field> fields = emptyList();
+
+    @ToString
+    @Getter
+    public static final class Field
+    {
+        /**
+         * A valid path can have up to 5 levels each with an alphanumeric name
+         * between 1 and 32 characters long and levels being separated by a dot.
+         *
+         * The path needs to be protected since it becomes part of the SQL when
+         * the path is extracted from the JSON values. Therefore, the
+         * limitations on the path are quite strict even if this will not allow
+         * some corner case names to be used that would be valid JSON member
+         * names.
+         */
+        private static final String PATH_PATTERN = "^[-_a-zA-Z0-9]{1,32}(?:\\.[-_a-zA-Z0-9]{1,32}){0,5}$";
+
+        private final String path;
+
+        private final String alias;
+
+        public Field( String path )
+        {
+            this( path, null );
+        }
+
+        public Field( String path, String alias )
+        {
+            if ( path == null || !path.matches( PATH_PATTERN ) )
+            {
+                throw new IllegalArgumentException( "Not a valid path: " + path );
```

Comment:
Wondering if we want to use an `ErrorCode` enum / exception here? Invalid paths might occur frequently and it might be good to have this error/conflict as part of the API docs.